### PR TITLE
Rewrite GitGraph.js

### LINKED ISSUE #11981: Damaged graph
- Can you reproduce the bug at https://try.gitea.io:
  - [x] Yes ([URL](https://try.gitea.io/ahaha/glu/graph))
  - [ ] No
  - [ ] Not relevant

## Description

We have damaged graph on some projects. This is very disturbing. Thanks.

### PR DESCRIPTION
**Pull request by zeripath**
The current vendored gitgraph.js is no longer maintained and is
difficult to understand, fix and maintain.

This PR completely rewrites its logic - hopefully in a clearer fashion
and easier to maintain.

It also includes @silverwind's improvements of coloring the commit dots
and preventing the flash of incorrect content.

Further changes to contemplate in future will be abstracting out of the
flows to an object, storing the involved commit references on the flows
etc. However, this is probably a required step for this.

Replaces #12131
Fixes #11981 (part 3)

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Comment by zeripath** [comment]
Although this is a rewrite it may be appropriate to backport to 1.12 as the gitgraph is broken there. Otherwise we could backport a simple hacky fix that will prevent the graph from disappearing by just assigning a '#000' flow whenever a flow is missing.

**Comment by zeripath** [comment]
Good example graphs to test this can be found in #11981:

https://try.gitea.io/ahaha/glu/graph
https://try.gitea.io/ahaha/qtconnectivity/graph
 https://try.gitea.io/ahaha/gitgraph/graph

File: go.mod [hunk 1]
```diff
@@ -100,8 +100,8 @@ require (
 	github.com/urfave/cli v1.20.0
 	github.com/xanzy/go-gitlab v0.31.0
 	github.com/yohcop/openid-go v1.0.0
-	github.com/yuin/goldmark-highlighting v0.0.0-20200307114337-60d527fdb691
 	github.com/yuin/goldmark v1.1.32
+	github.com/yuin/goldmark-highlighting v0.0.0-20200307114337-60d527fdb691
```
Comment (silverwind) [inline go.mod]: Seems unrelated.

File: web_src/js/features/gitgraph.js [hunk 2]
```diff
@@ -1,13 +1,579 @@
...
+    const devicePixelRatio = window.devicePixelRatio || 1;
+    const backingStoreRatio =
+      this.ctx.webkitBackingStorePixelRatio ||
+      this.ctx.mozBackingStorePixelRatio ||
+      this.ctx.msBackingStorePixelRatio ||
+      this.ctx.oBackingStorePixelRatio ||
+      this.ctx.backingStorePixelRatio ||
+      1;
+
+    const ratio = devicePixelRatio / backingStoreRatio;
```
Comment (silverwind) [inline web_src/js/features/gitgraph.js]:
```suggestion
    const ratio = window.devicePixelRatio || 1;
```

`backingStoreRatio` is deprecated and always `1` on all browsers so we can simplify this. See https://stackoverflow.com/questions/24332639

File: web_src/js/features/gitgraph.js [hunk 3]
```diff
@@ -1,13 +1,579 @@
...
+    i;
+  for (i = 0; i < stringLength; i++) {
+    rnum = Math.floor(Math.random() * chars.length);
+    randomString += chars.substring(rnum, rnum + 1);
+  }
+
+  return randomString;
+}
+
 export default async function initGitGraph() {
```
Comment (silverwind) [inline web_src/js/features/gitgraph.js:547]: Would prefer this to stay at the top of the file.

Comment (zeripath) [reply web_src/js/features/gitgraph.js:547]: Can't as eslint then moans that you're using the function before it's defined...

**Review by lafriks — APPROVED** [review]

**Review by silverwind — APPROVED** [review]

**Comment by lafriks** [comment]
:rocket:
